akka-javasdk/src/main/scala/akka/javasdk/impl/agent/GuardrailProvider.scala

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright (C) 2021-2025 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.javasdk.impl.agent
+
+import scala.concurrent.Future
+import scala.util.Failure
+
+import akka.actor.typed.ActorSystem
+import akka.annotation.InternalApi
+import akka.javasdk.agent.Guardrail
+import akka.javasdk.agent.GuardrailContext
+import akka.javasdk.agent.SimilarityGuard
+import akka.javasdk.impl.agent.ConfiguredGuardrail.UseFor
+import akka.runtime.sdk.spi.SpiAgent
+import com.typesafe.config.Config
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[javasdk] object GuardrailProvider {
+  final case class GuardrailEntry(configuredGuardrail: ConfiguredGuardrail, guardrail: Guardrail)
+  final case class AgentGuardrails(entries: Seq[GuardrailEntry]) {
+    private def collectGuardrails(useFor: UseFor): Seq[SpiAgent.Guardrail] =
+      entries.collect {
+        case entry if entry.configuredGuardrail.useFor.contains(useFor) => toSpiGuardrail(entry)
+      }
+
+    val modelRequestGuardrails: Seq[SpiAgent.Guardrail] =
+      collectGuardrails(UseFor.ModelRequest)
+    val modelResponseGuardrails: Seq[SpiAgent.Guardrail] =
+      collectGuardrails(UseFor.ModelResponse)
+    val mcpToolRequestGuardrails: Seq[SpiAgent.Guardrail] =
+      collectGuardrails(UseFor.McpToolRequest)
+    val mcpToolResponseGuardrails: Seq[SpiAgent.Guardrail] =
+      collectGuardrails(UseFor.McpToolResponse)
+  }
+
+  final class SpiGuardrailAdapter(entry: GuardrailEntry) extends SpiAgent.Guardrail {
+    private val guardrail = entry.guardrail
+
+    override def evaluate(text: String): Future[SpiAgent.Guardrail.Result] = {
+      val result = guardrail.evaluate(text)
+      Future.successful(new SpiAgent.Guardrail.Result(result.passed, result.explanation))
+    }
+
+    override val name: String =
+      entry.configuredGuardrail.name
+
+    override val category: String =
+      entry.configuredGuardrail.category
+
+    override val reportOnly: Boolean =
+      entry.configuredGuardrail.reportOnly
+  }
+
+  private def toSpiGuardrail(entry: GuardrailEntry): SpiAgent.Guardrail =
+    entry.guardrail match {
+      case g: SimilarityGuard => toSpiSimilarityGuard(g, entry.configuredGuardrail)
+      case _                  => new SpiGuardrailAdapter(entry)
+    }
+
+  private def toSpiSimilarityGuard(g: SimilarityGuard, c: ConfiguredGuardrail): SpiAgent.SimilarityGuard =
+    new SpiAgent.SimilarityGuard(c.name, c.category, c.reportOnly, g.badExamplesResourceDir, g.threshold)
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[javasdk] class GuardrailProvider(system: ActorSystem[_], applicationConfig: Config) {
```

Comment:
Not super important since internal, but, a few `final` here and there seems in place.